### ISSUE TITLE
TiC 2.2.2 Casting channel uncraftable

### ISSUE THREAD

**Issue by ypossem** [SlimeKnights/TinkersConstruct#2014]
I am unable to craft a casting channel with 5 seared bricks in the boat shape on either the top or bottom half of the crafting table. I am able to craft a faucet though.

tconstruct@1.9-2.2.2.jenkins156
FML@8.0.99.99
Forge@12.16.0.1863
mantle@1.9-0.9.4.jenkins114
mcp@9.19

**Comment by TherminatorX** [comment]
Channel is not currently in as of the 2.0 versions, despite how much people liked it.

**Comment by ypossem** [comment]
I couldn't find a changelog with this detail, so thanks for the heads up

**Comment by bonii-xx** [comment]
Casting channels will be back once somebody finds the time to implement a reliable version of them that don't lose liquids :L

**Comment by Robijnvogel** [comment]
Let me just dump these ideas here:
_Casting channel:

Has an internal inventory of x mB
Has a maximum acceptance rate of x mB per tick (per input)
Has an outflow rate of x mB per tick (per output)
If adjacent channel contains less (same fluid or empty), send small trickle to that channel.
Prioritise channel or container under it.
Channel can only receive from one source at a time. It will iterate through sources?
Channel can output to multiple outputs at once, but if it can output down, it will only output down.


The outputting channel will send an output request to all valid outputs
The accepting channel will determine what it will accept from and send the outputting channel a "message" telling it how much it will accept at once.
The outputting channel will then send over some liquids and waits for the accepting channel to send another acceptance message.
If the outputting channel is empty once it receives the message that it can send over some liquids, it will "send over" 0 mB instead.
Once the outputting channel is empty (or its level is equal to or lower than the container it was outputing to on the side), it will send an "I am empty message" to all containers it had previously sent a message
If a channel is accepting from multiple sources, it will send that it can receive (available liguid space)/(number of sources) from each of them. (this will possibly result in the channel not being completely filled, so a random one of those inputs should be allowed to output the residue mBs ((available liguid space)%(number of sources))

List<EnumFacing> adjacentOutputContainers; //max size would be 5, because a channel can't output to the top, unless another mod actively extracts liquid from this channel...
There should be distinguished between actively extracting and passively accepting containers
List<EnumFacing> containersCurrentlyAcceptingFrom; //sides from which this container has received a request to accept new liquids (and no cancel request)
List<EnumFacing> containersCurrentlyEjectingTo; //sides to which this container has sent a request to output liquids (and no cancel request)
Map<EnumFacing, Integer> mBCurrentlyAcceptingFrom //lists how many mBs this channel can still expect to receive from the container per side.

If a container block gets broken, its adjacent containers should be notified and empty their ArrayLists and Maps correspondingly
If a channel block gets placed, the adjacent containers should be registered
If a container block gets placed, the adjacent channels should register this_

Basically makes certain that liquid only streams to channels on an equal height or down-hill.
Works sort of like the internet. With requests, permission feedback, request cancelling, etc.
If I get time, I may be able to work on this. I realise that some of the functionality may already be part of Forge's standard Container, which should be expanded upon.

**Comment by Krands** [comment]
Hi!

So, to get this straight: casting channels are not in 1.10.2 for technical reasons?

**Comment by LovelyFlare** [comment]
Read the conversation...

**Comment by Robijnvogel** [comment]
@Krands, I think a combination of "technical issues" and "not enough time to look for a solution to those issues" would be more accurate.

**Comment by KnightMiner** [comment]
Resolved by #3213
